### ISSUE TITLE
last_called_summary does not update

### ISSUE THREAD

**Issue by Mis1ify** [alandtse/alexa_media_player#2275]
**IMPORTANT: Please search the issues, including closed issues, and the [FAQ](https://github.com/alandtse/alexa_media_player/wiki/FAQ) before opening a new issue. The template is mandatory; failure to use it will result in issue closure.**

**Describe the bug**
The 'last_called_summary' does not really update, or updates slowly.

This is an important state argument to me as I use it for automations. The problem started very long time ago. This happens only for 'last_called_summary'. For example, when I say 'volume up' the 'volume_up' argument raises, but 'last_called_summary' remains the same for a good one hour? Sometimes it will update within one minutes, but sometimes won't.

**To Reproduce**

1. Speak something to Alexa.
2. Check dev options of echo dot.
3. 'last_called_summary' was not updated.

**Expected behavior**
last_called_summary should update with max delay of 10-20 seconds.

**Screenshots**
![image](https://github.com/alandtse/alexa_media_player/assets/111151942/e1133f80-d121-4223-9059-0b273d559f5a)

**System details**

- Home-assistant (version): 2024.5.5
- alexa_media (version from `const.py` or HA startup): 4.10.0
- alexapy (version from `pip show alexapy` or HA startup): 1.27.10
- Amazon [2FA is enabled](https://github.com/alandtse/alexa_media_player/wiki/Configuration#enable-two-factor-authentication-for-your-amazon-account) (y/n). y

**Logs**
Please provide [logs](https://github.com/alandtse/alexa_media_player/wiki/FAQ#how-do-i-enable-debug-logging-for-the-component).

> 2024-06-04 10:45:24.308 WARNING (SyncWorker_0) [homeassistant.loader] We found a custom integration hacs which has not been tested by Home Assistant. This component might cause stability problems, be sure to disable it if you experience issues with Home Assistant
2024-06-04 10:45:24.315 WARNING (SyncWorker_0) [homeassistant.loader] We found a custom integration alexa_media which has not been tested by Home Assistant. This component might cause stability problems, be sure to disable it if you experience issues with Home Assistant
2024-06-04 10:45:34.456 WARNING (MainThread) [homeassistant.helpers.frame] Detected that custom integration 'hacs' accesses hass.components.frontend. This is deprecated and will stop working in Home Assistant 2024.9, it should be updated to import functions used from frontend directly at custom_components/hacs/frontend.py, line 68: hass.components.frontend.async_register_built_in_panel(, please create a bug report at https://github.com/hacs/integration/issues
2024-06-04 10:45:54.802 WARNING (MainThread) [homeassistant.helpers.frame] Detected that custom integration 'hacs' accesses hass.helpers.event. This is deprecated and will stop working in Home Assistant 2024.11, it should be updated to import functions used from event directly at custom_components/hacs/base.py, line 630: self.hass.helpers.event.async_track_time_interval(, please create a bug report at https://github.com/hacs/integration/issues
2024-06-04 10:46:38.037 DEBUG (MainThread) [alexapy.alexahttp2] Received raw message: Content-Type: application/json
{"directive":{"header":{"namespace":"Alexa.Mobile.Push","name":"RenderUpdate","messageId":"X"},"payload":{"renderingUpdates":[{"route":"EventBus:tcomm::message","resourceId":"PUSH_VOLUME_CHANGE","resourceMetadata":"{\"command\":\"PUSH_VOLUME_CHANGE\",\"timeStamp\":1717490797924,\"payload\":\"{\\\"dopplerId\\\":{\\\"deviceSerialNumber\\\":\\\"X\\\",\\\"deviceType\\\":\\\"A1RABVCI4QCIKC\\\"},\\\"volumeSetting\\\":100,\\\"isMuted\\\":false,\\\"destinationUserId\\\":\\\"X\\\"}\"}"}]}}}
--------abcde123
2024-06-04 10:46:38.040 DEBUG (MainThread) [custom_components.alexa_media] f**************i@g*******m: Received http2push command: PUSH_VOLUME_CHANGE : {'dopplerId': {'deviceSerialNumber': 'G************B0V', 'deviceType': 'A1RABVCI4QCIKC'}, 'volumeSetting': 100, 'isMuted': False, 'destinationUserId': 'A**********IPS'}
2024-06-04 10:46:38.040 DEBUG (MainThread) [custom_components.alexa_media] Adding PUSH_VOLUME_CHANGE to seen_commands: {}
2024-06-04 10:46:38.041 DEBUG (MainThread) [custom_components.alexa_media] Updating media_player volume: {'dopplerId': {'deviceSerialNumber': 'G************B0V', 'deviceType': 'A1RABVCI4QCIKC'}, 'volumeSetting': 100, 'isMuted': False, 'destinationUserId': 'A**********IPS'}
2024-06-04 10:46:38.042 DEBUG (MainThread) [custom_components.alexa_media.media_player] f**************i@g*******m: Fabian's Echo Dot volume updated: 100
2024-06-04 10:46:38.053 DEBUG (MainThread) [alexapy.alexahttp2] Received raw message: Content-Type: application/json
{"directive":{"header":{"namespace":"Alexa.Mobile.Push","name":"RenderUpdate","messageId":"X"},"payload":{"renderingUpdates":[{"route":"EventBus:tcomm::message","resourceId":"PUSH_EQUALIZER_STATE_CHANGE","resourceMetadata":"{\"command\":\"PUSH_EQUALIZER_STATE_CHANGE\",\"timeStamp\":1717490797920,\"payload\":\"{\\\"dopplerId\\\":{\\\"deviceSerialNumber\\\":\\\"X\\\",\\\"deviceType\\\":\\\"A1RABVCI4QCIKC\\\"},\\\"bass\\\":1,\\\"midrange\\\":-1,\\\"treble\\\":1,\\\"destinationUserId\\\":\\\"X\\\"}\"}"}]}}}
--------abcde123
2024-06-04 10:46:38.056 DEBUG (MainThread) [custom_components.alexa_media] f**************i@g*******m: Received http2push command: PUSH_EQUALIZER_STATE_CHANGE : {'dopplerId': {'deviceSerialNumber': 'G************B0V', 'deviceType': 'A1RABVCI4QCIKC'}, 'bass': 1, 'midrange': -1, 'treble': 1, 'destinationUserId': 'A**********IPS'}
2024-06-04 10:46:38.057 DEBUG (MainThread) [custom_components.alexa_media] Adding PUSH_EQUALIZER_STATE_CHANGE to seen_commands: {'PUSH_VOLUME_CHANGE': 1717490798.040848}
2024-06-04 10:46:38.057 DEBUG (MainThread) [custom_components.alexa_media] Updating media_player availability {'dopplerId': {'deviceSerialNumber': 'G************B0V', 'deviceType': 'A1RABVCI4QCIKC'}, 'bass': 1, 'midrange': -1, 'treble': 1, 'destinationUserId': 'A**********IPS'}
2024-06-04 10:46:41.187 DEBUG (MainThread) [alexapy.alexahttp2] Received raw message: Content-Type: application/json
{"directive":{"header":{"namespace":"Alexa.Mobile.Push","name":"RenderUpdate","messageId":"X"},"payload":{"renderingUpdates":[{"route":"EventBus:tcomm::message","resourceId":"PUSH_VOLUME_CHANGE","resourceMetadata":"{\"command\":\"PUSH_VOLUME_CHANGE\",\"timeStamp\":1717490801078,\"payload\":\"{\\\"dopplerId\\\":{\\\"deviceSerialNumber\\\":\\\"X\\\",\\\"deviceType\\\":\\\"X\\\"},\\\"volumeSetting\\\":90,\\\"isMuted\\\":false,\\\"destinationUserId\\\":\\\"X\\\"}\"}"}]}}}
--------abcde123
2024-06-04 10:46:41.189 DEBUG (MainThread) [custom_components.alexa_media] f**************i@g*******m: Received http2push command: PUSH_VOLUME_CHANGE : {'dopplerId': {'deviceSerialNumber': 'G************B0V', 'deviceType': 'A1RABVCI4QCIKC'}, 'volumeSetting': 90, 'isMuted': False, 'destinationUserId': 'A**********IPS'}
2024-06-04 10:46:41.190 DEBUG (MainThread) [custom_components.alexa_media] Updating media_player volume: {'dopplerId': {'deviceSerialNumber': 'G************B0V', 'deviceType': 'A1RABVCI4QCIKC'}, 'volumeSetting': 90, 'isMuted': False, 'destinationUserId': 'A**********IPS'}
2024-06-04 10:46:41.191 DEBUG (MainThread) [custom_components.alexa_media.media_player] f**************i@g*******m: Fabian's Echo Dot volume updated: 90

**Additional context**
Calling service 'alexa_media.update_last_called' updates the argument correctly, and i've even set up automation to run this every 10 seconds, but after some time it shows an error in logs TooManyRequests.

I understand that it may be a Alexa API limitation, but then why volume_level updates perfectly?

**Comment by alandtse** [comment]
Http2 streaming is not considered a separate request so is unaffected by 429 errors assuming it's allowed to connect. Your logs don't seem to show any use of last_called. But if the issue is that you're seeing 429s when trying to update, then you're out of luck since that's Amazon not responding so obviously we can't really figure out what was last called.

**Comment by Mis1ify** [comment]
The problem is I'm not seeing anything. The other settings are working correctly, and, as shown on logs, the volume level updated almost immediately, but before that I told something to Alexa, which is not apparently in the logs, so last_update was not updated.

**Comment by alandtse** [comment]
If you interacted with Alexa and it's not showing in the logs, it's possible Alexa removed the notification from http2 and now the function is broken permanently.

**Comment by Mis1ify** [comment]
It is working, after a good couple of minutes, instead of immediately. I've set up script to update last called every 10 seconds, and it works but the logs are filled of error with API: TooManyRequests.

**Comment by danielbrunt57** [comment]
> It is working, after a good couple of minutes, instead of immediately. I've set up script to update last called every 10 seconds, and it works but the logs are filled of error with API: To

That is an Amazon restriction!

I installed my integration with polling every 3600 seconds.  
The default 60 seconds gives me `Too many requests`.  
600 still gave me `Too many requests`.
3600 seems to be okay so far but it's far too long.  
So all of my automations/scripts update Last Alexa on an as needed basis before she speaks to the last Alexa spoken too.  I'll share my scripts if you like...

**Comment by Mis1ify** [comment]
That's what I thought, amazon restriction. It really sucks. Please, share your scripts if you may.

**Comment by danielbrunt57** [comment]
I found a post I had made in another issue...

Here is my `sensor.last_alexa`:

![image](https://github.com/custom-components/alexa_media_player/assets/49846893/ad057ccc-86b2-45d2-b022-d7bb5cce6929)
```
  - name: Last Alexa
    unique_id: last_alexa
    icon: phu:amazon-logo
    state: >
      {{ states.media_player | selectattr('attributes.last_called','eq',true) | map(attribute='entity_id') | first | default(none) }}
    availability: >
      {{ states.media_player | selectattr('attributes.last_called','eq',true) | map(attribute='attributes.last_called_timestamp') | first is defined }}
    attributes:
      last_called_timestamp: >
        {{ states.media_player | selectattr('attributes.last_called','eq',true) | map(attribute='attributes.last_called_timestamp') | first | default(none) }}
      actual_friendly_name: >
        {{ states.media_player | selectattr('attributes.last_called','eq',true) | map(attribute='attributes.friendly_name') | first | default(none) }}
```
I had also created a media_player.last_alexa via the Universal Media Player integration as something, somewhere, at one point required `entity_id: media_player.xxx` and did not accept templating:
```
media_player:
  - platform: universal
    name: "Last Alexa"
    unique_id: media_player.last_alexa
    state_template: "{{ states.media_player|selectattr('attributes.last_called','eq',true)|map(attribute='entity_id') | first | default(none) }}"
```

This is the script I invoke to update `sensor.last_alexa`.  Note the wait_for_trigger based on the attribute `last_called_timestamp`.
I added that attribute to sensor.last_alexa as waiting for the entity state to change did not work when the `from_state.state` and the `to_state.state` (i.e. media_player.bedroom_echo_dot_right) were unchanged but the `last_called_timestamp` had changed.

![image](https://github.com/custom-components/alexa_media_player/assets/49846893/d435bc67-c249-414c-bf39-78bad9b79c07)

```
alias: "Alexa: Update sensor.last_alexa - wait for tmestamp to update"
sequence:
  - parallel:
      - wait_for_trigger:
          - platform: state
            entity_id:
              - sensor.last_alexa
            attribute: last_called_timestamp
        timeout:
          hours: 0
          minutes: 0
          seconds: 5
          milliseconds: 0
      - service: alexa_media.update_last_called
        data: {}
    enabled: true
mode: single
```
(_when I look at my traces, the time remaining in the wait_for_trigger has been around 4.5 seconds meaning it took roughly 1/2 second to update_)

My other scripts always call the above script before doing anything else:
```
alias: Good Night
sequence:
  - service: script.update_last_alexa_and_wait
    data: {}
  - if:
      - condition: state
        entity_id: input_boolean.coffee_prepared
        state: "off"
    then:
      - service: script.activate_alexa_actionable_notification
        data:
          text: Did you prepare your coffee?
          event_id: actionable_notification_coffee_prepared
          alexa_device: "{{ states('media_player.last_alexa') }}"
  - if:
      - condition: state
        entity_id: input_boolean.garbage_is_out
        state: "off"
      - condition: or
        conditions:
          - condition: and
            conditions:
              - condition: state
                entity_id: sensor.garbage
                state: "0"
              - condition: time
                before: "09:00:00"
                weekday:
                  - wed
                after: "00:00:00"
          - condition: and
            conditions:
              - condition: state
                entity_id: sensor.garbage
                state: "1"
              - condition: time
                after: "12:00:00"
                weekday:
                  - tue
                before: "00:00:00"
    then:
      - service: script.activate_alexa_actionable_notification
        data:
          text: Did you put the garbage out?
          event_id: actionable_notification_garbage_pickup
          alexa_device: "{{ states('input_text.last_alexa') }}"
      - wait_for_trigger:
          - platform: event
            event_type: alexa_actionable_notification
            event_data:
              event_id: actionable_notification_garbage_pickup
        timeout:
          hours: 0
          minutes: 0
          seconds: 15
          milliseconds: 0
        enabled: false
  - if:
      - condition: state
        entity_id: input_boolean.recycling_is_out
        state: "off"
      - condition: or
        conditions:
          - condition: and
            conditions:
              - condition: state
                entity_id: sensor.recycling
                state: "0"
              - condition: time
                before: "09:00:00"
                weekday:
                  - thu
                  - fri
                after: "00:00:00"
          - condition: and
            conditions:
              - condition: state
                entity_id: sensor.recycling
                state: "1"
              - condition: time
                after: "12:01:00"
                weekday:
                  - wed
                  - thu
                before: "00:00:00"
    then:
      - service: script.activate_alexa_actionable_notification
        data:
          text: Did you put the recycling out?
          event_id: actionable_notification_recycling_pickup
          alexa_device: "{{ states('input_text.last_alexa') }}"
      - wait_for_trigger:
          - platform: event
            event_type: alexa_actionable_notification
            event_data:
              event_id: actionable_notification_recycling_pickup
        timeout:
          hours: 0
          minutes: 0
          seconds: 15
          milliseconds: 0
        enabled: false
  - parallel:
      - if:
          - condition: not
            conditions:
              - condition: state
                entity_id: climate.living_room
                attribute: preset_mode
                state: eco
                alias: Preset of Furnace
            alias: If Preset of Furnace is NOT  'eco'
          - condition: numeric_state
            entity_id: climate.living_room
            attribute: temperature
            above: 17
            alias: Furnace temperature is above 17
        then:
          - service: climate.set_temperature
            data:
              temperature: 17
            target:
              entity_id: climate.living_room
          - service: script.notify_alexa_media
            data:
              message: Heat is now set to 17
      - service: media_player.turn_off
        target:
          entity_id:
            - media_player.bedroom_tv
            - media_player.living_room_tv
        data: {}
      - if:
          - condition: state
            entity_id: lock.front_door_lock
            state: unlocked
        then:
          - service: lock.lock
            data: {}
            target:
              entity_id: lock.front_door_lock
          - service: script.notify_alexa_media
            data:
              message: The front door was unlocked but I locked it for you.
      - if:
          - condition: state
            entity_id: cover.garage_door
            state: open
        then:
          - if:
              - condition: state
                entity_id: sensor.season
                state: summer
            then:
              - service: script.notify_alexa_media
                data:
                  message: Hey. Did you know the garage door is open?
                  milliseconds: 0
            else:
              - service: cover.close_cover
                data:
                  entity_id: cover.garage_door
              - service: script.notify_alexa_media
                data:
                  message: I closed the garage door for you
                  milliseconds: 0
      - service: input_select.select_option
        data:
          option: Sleeping
        target:
          entity_id: input_select.sleep_status
      - service: input_boolean.turn_on
        data: {}
        target:
          entity_id:
            - input_boolean.sleep_mode
            - input_boolean.daniel_sleeping
      - service: input_number.set_value
        data:
          value: 0
        target:
          entity_id: input_number.bathroom_delay_off
      - service: light.turn_off
        data: {}
        target:
          area_id:
            - living_room
            - patio
      - if:
          - condition: numeric_state
            entity_id: sensor.fordpass_elveh
            below: 100
          - condition: state
            entity_id: sensor.wallbox_32a_charger
            state: "off"
        then:
          - service: script.notify_alexa_media
            data:
              message: >-
                Your Mach E battery range is only {{
                states("sensor.fordpass_elveh")|int(0) }} kilometers
  - delay:
      hours: 0
      minutes: 0
      seconds: 5
      milliseconds: 0
  - service: media_player.play_media
    continue_on_error: true
    data:
      media_content_id: Alexa.GoodNight.Play
      media_content_type: sequence
    target:
      entity_id: "{{ states('media_player.last_alexa') }}"
    alias: "Media Player: Say Goodnight"
  - wait_for_trigger:
      - platform: state
        entity_id:
          - sensor.last_alexa
        attribute: last_called_timestamp
    enabled: false
  - delay:
      hours: 0
      minutes: 0
      seconds: 3
      milliseconds: 0
  - service: light.turn_off
    target:
      entity_id:
        - light.bnl_l
        - light.bnl_r
        - light.bedroom_bedside_lamps
    data: {}
  - if:
      - condition: state
        entity_id: binary_sensor.bathroom_occupancy
        state: "off"
    then:
      - service: light.turn_off
        data: {}
        target:
          device_id: 93d77be359d4c8f809c2a09a581bea2e
mode: single
icon: hass:weather-night
```
I also issue all TTS via a single script which queues all utternaces with a variable delay based on the word count in the current message so they don't interrupt one another.  My `input_number.alexa_word_duration` is currently set to 0.5:
```
alias: "Alexa: Notify Alexa Media"
description: Send notification to sensor.last_alexa
fields:
  message:
    description: The message content
    example: I am notifying you
sequence:
  - service: notify.alexa_media
    continue_on_error: true
    data:
      message: "{{ message }}"
      data:
        type: tts
      target: "{{ states('media_player.last_alexa') }}"
  - delay:
      seconds: >-
        {{ (message.split(' ') | count *
        states('input_number.alexa_word_duration')|float ) | round(0, 'floor',
        default) }}
mode: queued
icon: mdi:account-voice
max: 10
```

**Comment by danielbrunt57** [comment]
> If you interacted with Alexa and it's not showing in the logs, it's possible Alexa removed the notification from http2 and now the function is broken permanently.

Mine was working a while ago (prior to June?) but definitely is not working now.
I'm not seeing `command == "PUSH_ACTIVITY"` in debug logs, or even a new `unknown` command.

@alandtse Please close this as unresolvable.

**Comment by danielbrunt57** [comment]
I might have a workaround for this but am still testing it on my system.
So far in the last 12 hours or so I have my last_called updating almost instantly and with no ill effects so far...

**Comment by danielbrunt57** [comment]
After two days of testing, my code modification seems to be working.  Do I have anyone interested in testing?

![image](https://github.com/user-attachments/assets/6af40990-cf5b-4ee7-9ca7-91e33dec8253)

Replace `__init__.py` with this modified version: [__init__.zip](https://github.com/user-attachments/files/16669546/__init__.zip)

**Comment by figorr** [comment]
Unfortunately It didn't work for me.

I am under AMP 4.12.10, OS 10.3.1 and Core 2024.8.2.

I deleted the picked file and replaced de `__init__.py` file with this modified version. Restarted HA ... and after a while ... the integration didn't work and the update_last_called is not being updated.

Even looks that the first time the last_called service (due to a manual call service, it is not mandatory to way to the normal pulling time) ... the integration fails and tts is not working until I reloaded the integration. Once I reloaded the integration the last_called is updated until the next call to the service.

**Comment by danielbrunt57** [comment]
I've just determined that method is unreliable, not sure why as it was working almost 100%.

I've switched to media_player.py and firing an event on volume change to HA bus with a restart-able automation to update last_called after a 3 second delay and initial tests are promising, but been there, done that. and failed so we'll see,..

New event fired on volume change:
![image](https://github.com/user-attachments/assets/dfd532d8-427f-4353-af89-b6517433ccd7)

New automation to listen for event and update last_called:
```
alias: Alexa Media - Update last_called
description: Updates last_called after a small delay following a volume change push from Amazon. Note: These volume change pushes occur from multiple devices when you say "Alexa", but also randomly for no reason, and also when the the answering Alexa responds, hence the difficulty in adapting this method of when to update last_called!
trigger:
  - platform: event
    event_type: alexa_media_volume_change
    id: alexa_media_player
condition: []
action:
  - delay:
      hours: 0
      minutes: 0
      seconds: 3
      milliseconds: 0
  - action: alexa_media.update_last_called
    data: {}
  - delay:
      hours: 0
      minutes: 0
      seconds: 1
      milliseconds: 0
  - event: alexa_last_called_updated
    event_data: {}
trace:
  stored_traces: 30
mode: restart
```
Your automations can then listen for the event: 'alexa_last_called_updated' before proceeding, or not!
